File: mobile/android/app/src/main/java/com/dots/herness/mobile/MobileGitClient.kt
package com.dots.herness.mobile

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.File

data class MobileGitResult(val code: Int, val output: String) {
    val succeeded get() = code == 0
    val unavailable get() = code == 2
}

private object NativeGit {
    private val loaded = runCatching { System.loadLibrary("herness_git") }.isSuccess

    @JvmStatic external fun availableNative(): Boolean
    @JvmStatic external fun executeNative(operation: String, repoPath: String, argument: String, second: String, token: String, authorName: String, authorEmail: String): String

    fun available(): Boolean = loaded && runCatching { availableNative() }.getOrDefault(false)
    fun execute(operation: String, repoPath: String, argument: String, second: String, token: String, authorName: String, authorEmail: String): MobileGitResult {
        if (!loaded) return MobileGitResult(2, "libgit2_not_linked")
        val raw = runCatching { executeNative(operation, repoPath, argument, second, token, authorName, authorEmail) }.getOrElse { "2:libgit2_not_linked" }
        val separator = raw.indexOf(':')
        return MobileGitResult(raw.substring(0, separator.coerceAtLeast(0)).toIntOrNull() ?: 1, raw.substring((separator + 1).coerceAtMost(raw.length)))
    }
}

/// Kotlin boundary over the C ABI. GitHub credentials are supplied in memory
/// for each operation and are never written to the repository configuration.
class MobileGitClient(private val root: () -> File, private val token: () -> String) {
    val available get() = NativeGit.available()

    suspend fun execute(operation: String, arguments: Map<String, String> = emptyMap()): MobileGitResult = withContext(Dispatchers.IO) {
        val workspace = root().canonicalFile
        val requestedPath = arguments["repoPath"]
        val repoPath = if (requestedPath == null) {
            workspace.path
        } else {
            val value = requestedPath
            val candidate = File(value).let { if (it.isAbsolute) it else File(workspace, value) }.canonicalFile
            if (candidate.path != workspace.path && !candidate.path.startsWith(workspace.path + File.separator)) {
                return@withContext MobileGitResult(1, "Git path is outside the workspace.")
            }
            candidate.path
        }
        NativeGit.execute(
            operation,
            repoPath,
            arguments["argument"].orEmpty(),
            arguments["second"].orEmpty(),
            token(),
            arguments["authorName"] ?: "HerNess Mobile",
            arguments["authorEmail"] ?: "mobile@herness.local",
        )
    }

    suspend fun clone(repositoryUrl: String, destination: File, branch: String) {
        val result = execute("clone", mapOf("repoPath" to destination.path, "argument" to repositoryUrl, "second" to branch))
        if (!result.succeeded) throw MobileGitException(result)
    }
}

class MobileGitException(private val result: MobileGitResult) : IllegalStateException(
    if (result.unavailable) "The native libgit2 backend is not linked in this build." else result.output.ifBlank { "The Git operation failed." }
)

File: mobile/ios/Sources/MobileGitClient.swift
import Foundation

struct MobileGitResult: Sendable {
    let code: Int32
    let output: String

    var succeeded: Bool { code == Int32(HERNESS_GIT_OK) }
    var unavailable: Bool { code == Int32(HERNESS_GIT_UNAVAILABLE) }
}

/// Small Swift boundary over the C ABI. The token is passed only for the
/// duration of the libgit2 call and is never put in a URL or Git config.
final class MobileGitClient: @unchecked Sendable {
    let root: URL
    private let token: @Sendable () -> String

    init(root: URL, token: @escaping @Sendable () -> String) {
        self.root = root
        self.token = token
    }

    var available: Bool { herness_git_available() != 0 }

    func execute(operation: String, arguments: [String: String] = [:]) async -> MobileGitResult {
        let root = self.root.resolvingSymlinksInPath().standardizedFileURL
        let repoPath: String
        if let value = arguments["repoPath"] {
            let candidate = (value.hasPrefix("/") ? URL(fileURLWithPath: value) : root.appendingPathComponent(value))
                .resolvingSymlinksInPath()
                .standardizedFileURL
            guard candidate.path == root.path || candidate.path.hasPrefix(root.path + "/") else {
                return MobileGitResult(code: Int32(HERNESS_GIT_ERROR), output: "Git path is outside the workspace.")
            }
            repoPath = candidate.path
        } else {
            repoPath = root.path
        }
        let token = token()
        let argument = arguments["argument"] ?? ""
        let second = arguments["second"] ?? ""
        let authorName = arguments["authorName"] ?? "HerNess Mobile"
        let authorEmail = arguments["authorEmail"] ?? "mobile@herness.local"
        return await Task.detached(priority: .userInitiated) {
            var buffer = [CChar](repeating: 0, count: 64 * 1024)
            let code = buffer.withUnsafeMutableBufferPointer { buffer in
                operation.withCString { operation in
                    repoPath.withCString { repoPath in
                        argument.withCString { argument in
                            second.withCString { second in
                                token.withCString { token in
                                    authorName.withCString { authorName in
                                        authorEmail.withCString { authorEmail in
                                            herness_git_execute(operation, repoPath, argument, second, token, authorName, authorEmail, buffer.baseAddress, buffer.count)
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
            }
            return MobileGitResult(code: code, output: String(cString: buffer))
        }.value
    }

    func clone(repositoryURL: String, destination: URL, branch: String) async throws {
        let result = await execute(operation: "clone", arguments: [
            "repoPath": destination.path,
            "argument": repositoryURL,
            "second": branch,
        ])
        guard result.succeeded else { throw MobileGitError(result: result) }
    }

    func fetch() async throws { try await require("fetch") }
    func checkout(branch: String) async throws { try await require("checkout", argument: branch) }
    func createBranch(_ branch: String) async throws { try await require("create_branch", argument: branch) }
    func commit(message: String, authorName: String, authorEmail: String) async throws {
        try await require("commit", argument: message, authorName: authorName, authorEmail: authorEmail)
    }
    func push(remote: String = "origin", branch: String = "HEAD") async throws {
        try await require("push", argument: remote, second: branch)
    }

    private func require(_ operation: String, argument: String = "", second: String = "", authorName: String = "HerNess Mobile", authorEmail: String = "mobile@herness.local") async throws {
        let result = await execute(operation: operation, arguments: [
            "argument": argument,
            "second": second,
            "authorName": authorName,
            "authorEmail": authorEmail,
        ])
        guard result.succeeded else { throw MobileGitError(result: result) }
    }
}

struct MobileGitError: LocalizedError, Sendable {
    let result: MobileGitResult

    var errorDescription: String? {
        result.unavailable
            ? "The native libgit2 backend is not linked in this build."
            : result.output.ifEmpty("The Git operation failed.")
    }
}

private extension String {
    func ifEmpty(_ fallback: String) -> String { isEmpty ? fallback : self }
}
